Destroy class hierarchy when destroying class with a `base`

Pixep opened the pull request:
Depends on #2361, #2421

Add support for destructors of base classes.
Features:
- Call destructors from derived to base class
- Support addressing `TupleValue` using a new `Member` variant

Changes:
- Update `StepDestroy()` to recursively call destructors from derived to base class
- Add new `Member` variant and `IndexedValue` struct to be usable with `TupleValue`

Pixep committed:
Explorer: support `.base` to initialize parent class from struct

- Add support for `.base` for parent class initialization
- Prevent parent class field initialization without using `.base`

Known limitations:
- Initialization of `base` field using a function or base constructor
  constructor is not supported

Pixep committed:
Merge branch 'trunk' into feature/base-class-init

Pixep committed:
testdata: add note for expected fail test with constructors

Pixep committed:
Explorer: separate NominalClassValue and base data

Add a new `base()` method to NominalClassValue to get its base class
object, including its fields.

Pixep committed:
Explorer: Remove dead code

Pixep committed:
Merge branch 'trunk' into feature/base-class-init

Pixep committed:
Explorer: rename 'FieldTypesWithParents'

... to 'FieldTypesWithBase', as suggested by @zygoloid.

Pixep committed:
Explorer: remove unused function declaration

jonmeow commented on explorer/interpreter/interpreter.cpp:
```diff
@@ -2034,14 +2087,21 @@ auto Interpreter::StepDestroy() -> ErrorOr<Success> {
-      if (class_dec.destructor().has_value()) {
-        return CallDestructor(*class_dec.destructor(), class_obj);
-      }
+  if (const auto* class_obj =
```
This seems weird how there's the tuple destruction in the middle, between this and 2130. Both of those are NominalClassValue handling; why are they split? Given you're removing the act.pos() == 0 check, I'm concerned this will start running logic out of order.

For more confidence in this, can you add the test:

```
package ExplorerTest api;

base class Data {
  fn Make(x: i32) -> Data { return {.data = x}; }
  destructor[me: Self]{
    Print("Destroying {0}", me.data);
  }
  var data: i32;
}

base class A {
  fn Make() -> B { return {.a1=Data.Make(8), .a2=Data.Make(7)}; }
  destructor[me: Self]{
    Print("Destroying 6 (A)", me.data);
  }
  var a1: Data;
  var a2: Data;
}

base class B extends A {
  fn Make() -> B { return {.base = A.Make(), .b1=Data.Make(5), .b2=Data.Make(4)}; }
  destructor[me: Self]{
    Print("Destroying 3 (B)", me.data);
  }
  var b1: Data;
  var b2: Data;
}

class C extends B {
  fn Make() -> C { return {.base = B.Make(), .c1=Data.Make(2), .c2=Data.Make(1)}; }
  destructor[me: Self]{
    Print("Destroying 0 (C)");
  }
  var c1: Data;
  var c2: Data;
}

fn Main() -> i32 {
  var c: C = C.Make();
  return 0;
}
```

I believe that should print:

Destroying 0 (C)
Destroying 1
Destroying 2
Destroying 3 (B)
Destroying 4
Destroying 5
Destroying 6 (A)
Destroying 7
Destroying 8

Pixep replied:
Haaa, you're totally right, thank you, that won't work as is. I didn't consider the rest of this fully.

Pixep replied:
Added

Charlesjourdain replied:
> # @~~````

````~~@hellotheworld quoi pestes sur le seul projet garder le donc y rien a gagner juste une folie etrangere .@

Pixep committed:
Merge branch 'trunk' into feature/base-class-init

Pixep committed:
Explorer: fix style, minor improvements

Pixep committed:
Explorer: add base class destructor support

Pixep committed:
Explorer: fix class member resolution error

Pixep committed:
Explorer: various improvements, generics wip

Pixep committed:
Explorer: refactor implicit conversion logic for classes

Pixep committed:
Explorer: removed unused include

Pixep committed:
Merge branch 'trunk' into feature/base-class-init

Pixep committed:
Explorer: add clarifying comment to NominalClassType::base_type()

Pixep committed:
Explorer: substitue members relative to their class

Fixes an issue causing base class members with generic types to be substituted

Pixep committed:
Merge branch 'trunk' into feature/base-class-init

Pixep committed:
Merge branch 'feature/base-class-init' into feature/destructor-base-class

Pixep committed:
Explorer: fix tuple destruction, add IndexableValue

Fix tuple destruction by providing a way to use `Member` for tuple
values, passing an `IndexableValue` (NamedValue, but with an index).

Pixep committed:
Explorer: support index-base fields and Members

Add `FieldPath::index() -> std::optional<size_t>`, a `Member`
variant for index-based, and `IndexedValue` mirroring `NamedValue`.

Pixep committed:
Explorer: add pre-commit changes

Pixep commented:
Created a dedicated PR for the FieldPath / Member changes for tuples https://github.com/carbon-language/carbon-lang/pull/2417. Will eventually drop from here, same as its dependent PR.

Pixep committed:
Explorer: improve Member API for named vs positional members

Given Member may or not be named, provide a better API for its usage:
- Member::Print(out)
- Member::IsNamed(std::string_view name) -> bool
- Member::isPositional() -> bool

Update dependent code to rely less on 'name()' by default

Pixep committed:
Explorer: use better names for getters

And refactor a bit

Pixep committed:
Explorer: enforce having a named member for `MemberName`

Pixep committed:
Explorer: Component API cleanup

Pixep committed:
Explorer: fix comment typo

Pixep committed:
Explorer: split Member into 2 classes

Split Member into (currently 2) classes, as we see the need for more Member variants
(base class access, possibly unammed mixins, ...), which in addition to
the current ones, also have significantly different attributes.

This will allow supporting more Member types in the feature in a clean
way.

Pixep committed:
Explorer: add BaseClass member

Needed for related work on base class destructor

Pixep committed:
Explorer: disambiguate base class output

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

Pixep committed:
Explorer: remove unecessary ClassDeclaration::base()

Pixep committed:
Explorer: remove superfluous CARBON_CHECKs

Pixep committed:
Merge branch 'feature/base-class-init' into feature/destructor-base-class

jonmeow reviewed (commented):
Although #2361 is in, I'm still seeing the changes from it when I look at the delta here. Can you try merging and re-pushing the branch?

jonmeow commented:
Note, I see a merge from the other feature branch -- I'm not sure if it'd help to update trunk on your fork, and merge from trunk

Pixep committed:
Merge branch 'trunk' into feature/destructor-base-class

Pixep commented:
With #2361 merged, this PR still has a soft dependency to https://github.com/carbon-language/carbon-lang/pull/2421 (and its https://github.com/carbon-language/carbon-lang/pull/2417 parent), which are tackling changes to `Member` separately, while also adding a way to get the base class using `Member`.

jonmeow commented on explorer/interpreter/interpreter.cpp:
```diff
@@ -1987,64 +1990,78 @@ auto Interpreter::StepDeclaration() -> ErrorOr<Success> {
         }
-        // Type of tuple element is integral type e.g. i32
-        // or the type has no destructor
+      } else if (act.pos() == member_count + 1) {
```
```suggestion
      } else if (act.pos() == member_count + 1) {
        // Destroy the parent, if there is one.
```

jonmeow commented on explorer/interpreter/interpreter.cpp:
```diff
@@ -1987,64 +1990,78 @@ auto Interpreter::StepDeclaration() -> ErrorOr<Success> {
-        if (item->kind() == Value::Kind::TupleValue) {
-          return todo_.Spawn(
-              std::make_unique<DestroyAction>(destroy_act.lvalue(), item));
+      } else if (act.pos() <= member_count) {
```
```suggestion
      } else if (act.pos() <= member_count) {
        // Destroy members.
```

jonmeow commented on explorer/interpreter/interpreter.cpp:
```diff
@@ -1987,64 +1990,78 @@ auto Interpreter::StepDeclaration() -> ErrorOr<Success> {
+      const auto& class_decl =
+          cast<NominalClassType>(class_obj->type()).declaration();
+      const auto member_count = static_cast<int>(class_decl.members().size());
+      if (act.pos() == 0) {
```
```suggestion
      if (act.pos() == 0) {
        // Run the destructor, if there is one.
```

jonmeow commented on explorer/interpreter/interpreter.cpp:
```diff
@@ -1987,64 +1990,78 @@ auto Interpreter::StepDeclaration() -> ErrorOr<Success> {
+          return todo_.Spawn(std::make_unique<DestroyAction>(
+              arena_->New<LValue>(field_address), item));
+        } else {
+          // Type of tuple element is integral type e.g. i32
+          // or the type has no destructor
```
```suggestion
          // The tuple element's type is an integral type (e.g., i32)
          // or the type doesn't support destruction.
```

Trying to fix punctuation but also clarify a little: destructors are optional on classes, but a class without a destructor still needs DestroyAction run on it.

This does make me wonder if some kind of is_destructible() attribute might add some helpful guard rails, but that's just me musing, probably not worth doing right now.

Pixep replied:
Yeah it might be... and to be fair I kept this comment as-is from existing code, and haven't delved into what was appropriate or not in that regard. On the surface I would tend to agree with your comment.
Done for the comment changes.

jonmeow commented on explorer/interpreter/interpreter.cpp:
```diff
@@ -1987,64 +1990,78 @@ auto Interpreter::StepDeclaration() -> ErrorOr<Success> {
+      const auto* class_obj = cast<NominalClassValue>(destroy_act.value());
+      const auto& class_decl =
+          cast<NominalClassType>(class_obj->type()).declaration();
+      const auto member_count = static_cast<int>(class_decl.members().size());
```
```suggestion
      const int member_count = class_decl.members().size();
```

Does this work without a cast?

Pixep replied:
Good point, yes.

Pixep committed:
Merge branch 'trunk' into feature/tuple-member-variant

jonmeow commented on explorer/ast/member.h:
```diff
@@ -25,6 +25,16 @@ struct NamedValue {
   Nonnull<const Value*> value;
 };
 
+// A IndexedValue represents a value identified by an index, such as a tuple
+// field
+struct IndexedValue {
+  // The field index.
+  size_t index;
+
+  // The field's value.
+  Nonnull<const Value*> value;
+};
```
Looking at how other code handles this, I see NamedValue used:
https://github.com/carbon-language/carbon-lang/blob/trunk/explorer/interpreter/interpreter.cpp#L468-L473

That makes this feel a bit like it's adding a new way to index tuples; can we instead do it with NamedValue per the referenced code? I believe then member also wouldn't be changed, since NamedValue is already supported.

(note, I'm omitting a few comments specific to IndexedValue assuming it'll be removed)

Pixep committed:
Merge branch 'feature/tuple-member-variant' into feature/member-classes

Pixep committed:
Explorer: fix method naming & int type preference

Pixep committed:
Explorer: make "Member::IsNamed" virtual, minor fixes

Pixep committed:
Explorer: specialize GetMember* functions and params

Pixep commented:
Depends on https://github.com/carbon-language/carbon-lang/pull/2421
(Restore dependency needed for clean merge)

Pixep committed:
Explorer: add comments for destruction steps

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

Pixep committed:
Explorer: member_test: setup dependencies properly

Pixep committed:
Explorer: rename `*Member` classes to `*Element`

Rename related code to use the same naming

Pixep committed:
Merge branch 'feature/tuple-member-variant' into feature/member-classes

Pixep committed:
Merge branch 'trunk' into feature/member-classes

Pixep committed:
Explorer: add missing change

Pixep committed:
Explorer: add new pre-commit changes

Pixep committed:
Explorer: Fix `BaseElement` returning type instead of value

Pixep committed:
Explorer: rename 'Member/TupleElement' to 'Named/Positional'

Pixep committed:
Explorer: rename `GetMember` to `GetElement`

Pixep committed:
Merge branch 'feature/member-classes' into feature/destructor-base-class

Pixep committed:
Explorer: replace 'self' with 'me'

Pixep committed:
Merge branch 'trunk' into feature/destructor-base-class

Pixep committed:
Explorer: minor simplifications

jonmeow reviewed (approved):
Thanks! Looks good, especially with the dependencies merged. ;)
